src/basic/virt.c

```diff
@@ -485,6 +485,51 @@ int detect_virtualization(void) {
         return r;
 }
 
+int running_in_userns(void) {
+        _cleanup_fclose_ FILE *f = NULL;
+        _cleanup_free_ char *buf = NULL;
+        size_t n_allocated = 0;
+        ssize_t n;
+        uint32_t a, b, c;
+        int r;
+
+        f = fopen("/proc/self/uid_map", "re");
+        if (!f) {
+                log_debug_errno(errno, "Failed to open /proc/self/uid_map: %m");
+                return errno == -ENOENT ? false : -errno;
+        }
+
+        n = getline(&buf, &n_allocated, f);
+        if (n < 0) {
+                if (feof(f)) {
+                        log_debug("/proc/self/uid_map is empty, user namespace is uninitialized");
+                        return true;
+                }
+
+                return log_debug_errno(errno, "Failed to read /proc/self/uid_map: %m");
+        }
+
+        r = sscanf(buf, "%"PRIu32" %"PRIu32" %"PRIu32, &a, &b, &c);
+        if (r < 3)
+                return log_debug_errno(errno, "Failed to parse /proc/self/uid_map: %m");
+
+        if (a > 0 || b > 0 || c != UINT32_MAX) {
+                log_debug("/proc/self/uid_map has mapping: %"PRIu32" %"PRIu32" %"PRIu32, a, b, c);
+                return true;
+        }
+
+        n = getline(&buf, &n_allocated, f);
+        if (feof(f)) {
+                log_debug("/proc/self/uid_map has a full 1:1 mapping");
```

Comment:
Hmm,  not sure I grok this log message? Why do you say it has a full 1:1 mapping? Having a single line only means that you have one mapping only, notthing more... Or am I missing something here?